pkg/apis/fleet.cattle.io/v1alpha1/generic_map.go

```diff
@@ -0,0 +1,57 @@
+package v1alpha1
+
+import (
+	"encoding/json"
+)
+
+type GenericMap struct {
+	Data map[string]interface{} `json:"-"`
+}
+
+func (in GenericMap) MarshalJSON() ([]byte, error) {
+	return json.Marshal(in.Data)
+}
+
+func (in *GenericMap) UnmarshalJSON(data []byte) error {
+	in.Data = map[string]interface{}{}
+	return json.Unmarshal(data, &in.Data)
+}
+
+func (in *GenericMap) DeepCopyInto(out *GenericMap) {
+	out.Data = make(map[string]interface{}, len(in.Data))
+	deepCopyMap(in.Data, out.Data)
+}
+
+func deepCopyMap(src, dest map[string]interface{}) {
+	for key := range src {
+		switch value := src[key].(type) {
+		case map[string]interface{}:
+			destValue := make(map[string]interface{}, len(value))
+			deepCopyMap(value, destValue)
+			dest[key] = destValue
+		case []any:
+			destValue := make([]any, len(value))
+			deepCopySlice(value, destValue)
+			dest[key] = destValue
+		default:
```

Comment:
Can pointer types ever come as input?


Eg. pointers to slices or maps.

Reply:
These changes rely on `GenericMap` always being instantiated from calling Unmarshal over a JSON or YAML payload, which should not produce such pointers. Note that this code also won't work for for slices and maps with a specific type (e.g. `map[string]string` vs. `map[string]interface{}`). That's the main reason I kept this function here instead of a common place where it could be reused.

In fact, I should probably add this as a comment so that we can prevent blindly copy-pasting it somewhere that may not fit that restriction. Thanks for your comment!

Reply:
Comments added at b9ddf4a437fae3895494b5e68c4be5d8ea671b87